libzecale/circuits/fields/fp6_3over2_gadgets.tcc

```diff
@@ -0,0 +1,214 @@
+    , _c2(pb, FMT(annotation_prefix, " c2"))
+{
+}
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T>::Fp6_3over2_variable(
+    libsnark::protoboard<FieldT> &pb,
+    const Fp6_3over2_variable<Fp6T> &v,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _c0(v._c0)
+    , _c1(v._c1)
+    , _c2(v._c2)
+{
+}
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T>::Fp6_3over2_variable(
+    libsnark::protoboard<FieldT> &pb,
+    const Fp6T &v,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _c0(pb, v.c0, FMT(annotation_prefix, " c0"))
+    , _c1(pb, v.c1, FMT(annotation_prefix, " c1"))
+    , _c2(pb, v.c2, FMT(annotation_prefix, " c2"))
+{
+}
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T>::Fp6_3over2_variable(
+    libsnark::protoboard<FieldT> &pb,
+    const libsnark::Fp2_variable<Fp2T> &c0,
+    const libsnark::Fp2_variable<Fp2T> &c1,
+    const libsnark::Fp2_variable<Fp2T> &c2,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix), _c0(c0), _c1(c1), _c2(c2)
+{
+}
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T> Fp6_3over2_variable<Fp6T>::operator*(
+    const FieldT &scalar)
+{
+    return Fp6_3over2_variable<Fp6T>(
+        this->pb,
+        _c0 * scalar,
+        _c1 * scalar,
+        _c2 * scalar,
+        FMT(this->annotation_prefix, " *Fp"));
+}
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T> Fp6_3over2_variable<Fp6T>::operator+(
+    const Fp6_3over2_variable<Fp6T> &other)
+{
+    return Fp6_3over2_variable<Fp6T>(
+        this->pb,
+        _c0 + other._c0,
+        _c1 + other._c1,
+        _c2 + other._c2,
+        FMT(this->annotation_prefix.c_str(),
+            " + %s",
+            other.annotation_prefix.c_str()));
+}
+
+template<typename Fp6T> void Fp6_3over2_variable<Fp6T>::evaluate() const
+{
+    _c0.evaluate();
+    _c1.evaluate();
+    _c2.evaluate();
+}
+
+template<typename Fp6T>
+void Fp6_3over2_variable<Fp6T>::generate_r1cs_witness(const Fp6T &v)
+{
+    _c0.generate_r1cs_witness(v.c0);
+    _c1.generate_r1cs_witness(v.c1);
+    _c2.generate_r1cs_witness(v.c2);
+}
+
+template<typename Fp6T> Fp6T Fp6_3over2_variable<Fp6T>::get_element() const
+{
+    return Fp6T(_c0.get_element(), _c1.get_element(), _c2.get_element());
+}
+
+// Fp6_3over2_mul_gadget methods
+
+template<typename Fp6T>
+Fp6_3over2_mul_gadget<Fp6T>::Fp6_3over2_mul_gadget(
+    libsnark::protoboard<FieldT> &pb,
+    const Fp6_3over2_variable<Fp6T> &A,
+    const Fp6_3over2_variable<Fp6T> &B,
+    const Fp6_3over2_variable<Fp6T> &result,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _A(A)
+    , _B(B)
+    , _result(result)
+    , _v1(pb,
+          A._c1,
+          B._c1,
+          libsnark::Fp2_variable<Fp2T>(pb, FMT(annotation_prefix, " a1*b1")),
+          FMT(annotation_prefix, " _a1b1"))
```

Comment:
Why not rename the annotation as `_v1` here? (same below) so that the "annotation" and the "variable"/"wire name" are the same.